File: app/api/routes/documents.py
"""Document intake: multipart upload, canonical Postgres row, storage, enqueue."""

from __future__ import annotations

import uuid
from typing import Annotated
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, Response, UploadFile
from fastapi.responses import RedirectResponse
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import get_db, get_object_storage_dep
from app.auth.dependencies import get_current_user
from app.core.config import get_settings
from app.db.models import DocumentScore
from app.schemas.document import (
    CanonicalScoreOut,
    DocumentDetailOut,
    DocumentListResponse,
    DocumentSourceOut,
    DocumentSummaryOut,
    IntakeResponse,
    UrlIntakeRequest,
    UrlIntakeResponse,
)
from app.schemas.pipeline import DocumentPipelineOut
from app.services.document_service import (
    delete_document_for_user,
    get_document_for_user,
    list_documents_for_user,
    run_file_intake,
    run_url_intake_submit,
)
from app.services.exceptions import IntakeValidationError, StorageUploadError
from app.services.pipeline_status_service import get_document_pipeline_for_user
from app.services.storage_service import ObjectStorage
from app.services.user_metadata import parse_metadata_json_string, validate_user_metadata

router = APIRouter(prefix="/documents", tags=["documents"])


@router.get("", response_model=DocumentListResponse)
def list_documents(
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user),
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> DocumentListResponse:
    """
    List documents in collections the caller can access (org membership; optional dev fallback).
    """
    items, total = list_documents_for_user(db, auth_sub=user_id, limit=limit, offset=offset)
    return DocumentListResponse(
        items=[DocumentSummaryOut.model_validate(d) for d in items],
        total=total,
        user_id=user_id,
    )


@router.post("/from-url", response_model=UrlIntakeResponse, status_code=202)
def ingest_document_from_url(
    body: UrlIntakeRequest,
    db: Session = Depends(get_db),
    _user_id: str = Depends(get_current_user),
) -> UrlIntakeResponse:
    """
    Accept a remote URL: `created` row + `url` source, then worker fetch → S3 → pipeline.

    Poll `GET /documents/{id}` for `queued` / `failed` after the fetch job completes.
    """
    _ = _user_id
    try:
        payload = run_url_intake_submit(
            db,
            raw_url=body.url,
            collection_id_param=body.collection_id,
            title=body.title,
            user_metadata=body.metadata,
        )
    except IntakeValidationError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    return UrlIntakeResponse(**payload)


@router.get("/{document_id}/pipeline", response_model=DocumentPipelineOut)
def get_document_pipeline(
    document_id: uuid.UUID,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user),
) -> DocumentPipelineOut:
    """Latest pipeline run and events for polling + UI progress (worker writes to Postgres)."""
    out = get_document_pipeline_for_user(db, document_id=document_id, auth_sub=user_id)
    if out is None:
        raise HTTPException(status_code=404, detail="Document not found")
    return out


def _content_disposition_attachment(filename: str) -> str:
    ascii_fallback = filename.encode("ascii", "ignore").decode("ascii").strip() or "download.bin"
    ascii_fallback = ascii_fallback.replace('"', "").replace("\\", "")
    utf8_quoted = quote(filename, safe="")
    return f"attachment; filename=\"{ascii_fallback}\"; filename*=UTF-8''{utf8_quoted}"


@router.get(
    "/{document_id}/file",
    response_model=None,
    responses={
        302: {"description": "Redirect to a short-lived presigned object URL (when supported)."},
        404: {"description": "Document not found, no original on file, or object missing."},
        502: {"description": "Storage read failure when streaming the body."},
    },
)
def download_document_original(
    document_id: uuid.UUID,
    db: Session = Depends(get_db),
    storage: ObjectStorage = Depends(get_object_storage_dep),
    _user_id: str = Depends(get_current_user),
    redirect: Annotated[
        bool,
        Query(
            description=(
                "When true (default), return 302 to a presigned GET URL if the storage backend "
                "supports signing. Otherwise stream bytes through this API "
                "(still requires Bearer auth)."
            ),
        ),
    ] = True,
) -> Response | RedirectResponse:
    """
    Download the stored original bytes (same access rules as GET /documents/{id}).

    With real S3/MinIO, **redirect=true** avoids proxying large files through the API. Clients that
    cannot follow redirects to the object host (or need a single authenticated hop) should use
    **redirect=false**.
    """
    settings = get_settings()
    out = get_document_for_user(db, document_id=document_id, auth_sub=_user_id)
    if out is None:
        raise HTTPException(status_code=404, detail="Document not found")
    doc, _sources = out
    if not doc.storage_key:
        raise HTTPException(
            status_code=404,
            detail="Original file is not available for this document",
        )
    if not storage.object_exists(doc.storage_key):
        raise HTTPException(
            status_code=404,
            detail="Original file is missing from object storage",
        )

    filename = doc.original_filename or doc.storage_key.rsplit("/", 1)[-1]
    media_type = doc.content_type or "application/octet-stream"

    if redirect:
        signed = storage.presigned_get_url(
            doc.storage_key,
            expires_seconds=settings.download_presigned_ttl_seconds,
        )
        if signed:
            return RedirectResponse(url=signed, status_code=302)

    try:
        body = storage.get_bytes(doc.storage_key)
    except KeyError:
        raise HTTPException(
            status_code=404,
            detail="Original file is missing from object storage",
        ) from None
    except StorageUploadError as e:
        raise HTTPException(
            status_code=502,
            detail={"message": "failed to read object from storage", "error": str(e)},
        ) from e

    return Response(
        content=body,
        media_type=media_type,
        headers={"Content-Disposition": _content_disposition_attachment(filename)},
    )


@router.get("/{document_id}", response_model=DocumentDetailOut)
def get_document(
    document_id: uuid.UUID,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user),
) -> DocumentDetailOut:
    """Single document with intake sources (locators, mime, sizes)."""
    out = get_document_for_user(db, document_id=document_id, auth_sub=user_id)
    if out is None:
        raise HTTPException(status_code=404, detail="Document not found")
    doc, sources = out
    base = DocumentSummaryOut.model_validate(doc)
    score_row = db.scalar(
        select(DocumentScore)
        .where(
            DocumentScore.document_id == doc.id,
            DocumentScore.is_canonical.is_(True),
        )
        .limit(1)
    )
    canon: CanonicalScoreOut | None = None
    if score_row is not None:
        canon = CanonicalScoreOut(
            factuality_score=float(score_row.factuality_score)
            if score_row.factuality_score is not None
            else None,
            ai_generation_probability=float(score_row.ai_generation_probability)
            if score_row.ai_generation_probability is not None
            else None,
            fallacy_score=float(score_row.fallacy_score)
            if score_row.fallacy_score is not None
            else None,
            confidence_score=float(score_row.confidence_score)
            if score_row.confidence_score is not None
            else None,
            scorer_name=score_row.scorer_name,
            scorer_version=score_row.scorer_version,
        )
    return DocumentDetailOut(
        **base.model_dump(),
        sources=[DocumentSourceOut.model_validate(s) for s in sources],
        body_text=doc.body_text,
        canonical_score=canon,
    )


@router.delete("/{document_id}", status_code=204)
def remove_document(
    document_id: uuid.UUID,
    db: Session = Depends(get_db),
    storage: ObjectStorage = Depends(get_object_storage_dep),
    user_id: str = Depends(get_current_user),
) -> Response:
    """Delete canonical row (cascades sources) and remove raw object from storage when possible."""
    ok = delete_document_for_user(
        db,
        document_id=document_id,
        auth_sub=user_id,
        storage=storage,
    )
    if not ok:
        raise HTTPException(status_code=404, detail="Document not found")
    return Response(status_code=204)


@router.post("", response_model=IntakeResponse)
def upload_document(
    file: UploadFile = File(..., description="Raw file bytes for intake"),
    collection_id: str | None = Form(
        default=None,
        description=(
            "Target collection UUID; defaults to VERIFIEDSIGNAL_DEFAULT_COLLECTION_ID when omitted"
        ),
    ),
    title: str | None = Form(
        default=None,
        description="Optional display title (defaults to filename)",
    ),
    metadata: str | None = Form(
        default=None,
        description='Optional JSON object, e.g. {"tags":["finance"],"label":"report"}',
    ),
    db: Session = Depends(get_db),
    storage: ObjectStorage = Depends(get_object_storage_dep),
    _user_id: str = Depends(get_current_user),
) -> IntakeResponse:
    """
    Phase 1 intake: validate, insert canonical `documents` row (`created`), upload to S3/MinIO,
    finalize to `queued` with `document_sources`, enqueue `process_document`.

    Sync handler so `asyncio.run` inside storage/queue helpers is safe (runs on a worker thread).
    """
    raw = file.file.read()
    try:
        um = validate_user_metadata(parse_metadata_json_string(metadata))
        payload = run_file_intake(
            db,
            file_bytes=raw,
            original_filename=file.filename or "upload",
            content_type=file.content_type,
            title=title,
            collection_id_param=collection_id,
            user_metadata=um,
            storage=storage,
            auth_sub=_user_id,
        )
    except IntakeValidationError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except StorageUploadError as e:
        did = e.document_id
        raise HTTPException(
            status_code=502,
            detail={
                "message": "object storage upload failed",
                "error": str(e),
                "document_id": str(did) if did is not None else None,
            },
        ) from e

    return IntakeResponse(**payload)

File: apps/web/src/api/documents.ts
import { getApiBaseUrl } from "../config";
import { ApiError, apiFetch, joinApiUrl, readErrorMessage } from "./http";
import type { DocumentDetail, DocumentListResponse, IntakeResponse, UrlIntakeResponse } from "./types";

export async function listDocuments(
  accessToken: string,
  params: { limit?: number; offset?: number } = {},
): Promise<DocumentListResponse> {
  const sp = new URLSearchParams();
  if (params.limit != null) sp.set("limit", String(params.limit));
  if (params.offset != null) sp.set("offset", String(params.offset));
  const q = sp.toString();
  const path = q ? `/api/v1/documents?${q}` : "/api/v1/documents";
  const res = await apiFetch(path, { accessToken });
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
  return (await res.json()) as DocumentListResponse;
}

export async function getDocument(accessToken: string, documentId: string): Promise<DocumentDetail> {
  const res = await apiFetch(`/api/v1/documents/${encodeURIComponent(documentId)}`, { accessToken });
  if (res.status === 404) {
    throw new ApiError("Document not found", 404);
  }
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
  return (await res.json()) as DocumentDetail;
}

export async function uploadDocumentFile(
  accessToken: string,
  file: File,
  options: { collectionId?: string; title?: string } = {},
): Promise<IntakeResponse> {
  const fd = new FormData();
  fd.append("file", file, file.name);
  if (options.collectionId) fd.append("collection_id", options.collectionId);
  if (options.title) fd.append("title", options.title);
  const res = await apiFetch("/api/v1/documents", {
    method: "POST",
    accessToken,
    body: fd,
    isFormData: true,
  });
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
  return (await res.json()) as IntakeResponse;
}

export async function ingestDocumentFromUrl(
  accessToken: string,
  payload: { url: string; collection_id?: string | null; title?: string | null },
): Promise<UrlIntakeResponse> {
  const res = await apiFetch("/api/v1/documents/from-url", {
    method: "POST",
    accessToken,
    jsonBody: payload,
  });
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
  return (await res.json()) as UrlIntakeResponse;
}

export async function deleteDocument(accessToken: string, documentId: string): Promise<void> {
  const res = await apiFetch(`/api/v1/documents/${encodeURIComponent(documentId)}`, {
    method: "DELETE",
    accessToken,
  });
  if (res.status === 404) {
    throw new ApiError("Document not found", 404);
  }
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
}

/**
 * Streams the original file through the API (`redirect=false`) so the Bearer token is sent.
 * Triggers a browser download with the filename from Content-Disposition when present.
 */
export async function downloadOriginalFile(accessToken: string, documentId: string): Promise<void> {
  const root = getApiBaseUrl();
  if (!root) {
    throw new ApiError("VITE_API_URL is not set", 0);
  }
  const path = `/api/v1/documents/${encodeURIComponent(documentId)}/file?redirect=false`;
  const url = joinApiUrl(root, path);
  const res = await fetch(url, {
    method: "GET",
    headers: { Authorization: `Bearer ${accessToken}` },
    credentials: "include",
  });
  if (res.status === 404) {
    throw new ApiError("Original file not available", 404);
  }
  if (!res.ok) {
    throw new ApiError(await readErrorMessage(res), res.status);
  }
  const blob = await res.blob();
  const cd = res.headers.get("Content-Disposition");
  let filename = "download";
  if (cd) {
    const m = /filename\*=UTF-8''([^;]+)|filename="([^"]+)"/i.exec(cd);
    const raw = m?.[1] || m?.[2];
    if (raw) {
      try {
        filename = decodeURIComponent(raw);
      } catch {
        filename = raw;
      }
    }
  }
  const href = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = href;
  a.download = filename;
  a.rel = "noopener";
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(href);
}
